kernel/EigenSolvers/JacobiDavidsonMaxwell.h

```diff
@@ -0,0 +1,144 @@
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#ifndef JACOBIDAVIDSONMAXWELLSOLVER_H
+#define JACOBIDAVIDSONMAXWELLSOLVER_H
+
+#include <petsc.h>
+#include <petscksp.h>
+
+#include <slepcbv.h>
+#include <slepceps.h>
+
+#include <iostream>
+#include <math.h>
+#include <cstdlib>
+#include <vector>
+
+#include <chrono>  // For timing
+
+#include <utility>
+#include <valarray>
+
+/**
+ * Solver of the Maxwell equation based on the paper
+ *   Multilevel preconditioned iterative eigensolvers
+ *   for Maxwell eigenvalue problems
+ *   P Arbenz R Geus
+ *   Applied Numerical Mathematics 54 (2005) 107-121
+ *
+ *
+ * Implementation note:
+ *
+ */
+
+namespace hpgem {
+
+namespace EigenSolvers {
+
+class JacobiDavidsonMaxwellSolver final {
+
+   public:
+    JacobiDavidsonMaxwellSolver();
+
+    void setMatrices(const Mat Ain, const Mat Cin);
+    PetscErrorCode solve(PetscInt nev);
+    PetscInt getConverged();
+    PetscErrorCode getEigenPair(PetscInt index, PetscScalar &eval, Vec &evec);
+    PetscInt getIterationCount();
+
+    void setMaxIter(int n);
+    void setSearchSpaceMaxSize(int n);
+    void setCorrectionNiter(int n);
+    void setTolerance(PetscReal tol);
+
+   private:
+    void initializeMatrices();
+    void initializeVectors();
+    void initializeSearchSpace(int nev);
+
+    PetscErrorCode computeRayleighQuotient(const Vec &x, PetscReal *out);
+    PetscErrorCode normalizeVector(Vec &x);
+    PetscErrorCode VecxTAx(const Vec &x, const Mat &K, PetscScalar *val);
+    PetscErrorCode getCorrectionOperator(Mat &op);
+    PetscErrorCode solveCorrectionEquation(const Vec &res, Vec &sol);
+    PetscErrorCode computeResidueVector(const Vec &q, const PetscReal rho,
+                                        Vec &res);
+
+    PetscErrorCode computeProjection(Vec &v, const BV &Q);
+
+    PetscErrorCode projectCorrectionVector(Vec &corr);
+    PetscErrorCode addVectorToSearchSpace(const Vec &v, PetscInt idx);
+    PetscErrorCode computeSmallEigenvalues(std::vector<PetscReal> &eval,
+                                           Vec *evec);
+    PetscErrorCode correctionOperatorMatMult(Vec x, Vec y);
+    static PetscErrorCode staticMatMultCorrOp(Mat M, Vec x, Vec y);
+    PetscErrorCode computeThreshold(Vec q, Vec r, PetscReal *eps);
+    PetscErrorCode correctionPreconditionerMatMult(Vec x, Vec y);
+    static PetscErrorCode staticMatMultCorrPrec(Mat M, Vec x, Vec y);
+
+    PetscReal tau = 1.2;
+    PetscReal eta;
+    PetscInt maxIter = 100;
+    PetscInt correction_niter = 10;
+    PetscInt iter = 0;
+    PetscInt search_space_maxsize = 25;
+    PetscInt search_space_minsize;
+    PetscInt search_space_current_size = 0;
+    PetscInt V_current_size = 0;
+    PetscInt Q_current_size = 0;
+    PetscInt Qt_current_size = 0;
+    PetscInt nconverged = 0;
+    PetscReal tolerance = 1E-3;
+
+    Mat A, C;
+    Mat Y, H;
+    BV Qt, Q, V;
```

Comment:
As noted above, I've got no clue what the difference is between Qt and Q. Could you elaborate?